src/Symfony/Component/Security/Core/User/LdapUserProvider.php

```diff
@@ -0,0 +1,105 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\User;
+
+use Symfony\Component\Security\Core\Exception\UnsupportedUserException;
+use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+use Symfony\Component\Security\Ldap\Exception\ConnectionException;
+use Symfony\Component\Security\Ldap\LdapInterface;
+
+/**
+ * LdapUserProvider is a simple user provider on top of ldap.
+ *
+ * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ * @author Charles Sarrazin <charles@sarraz.in>
+ */
+class LdapUserProvider implements UserProviderInterface
+{
+    private $ldap;
+    private $baseDn;
+    private $searchDn;
+    private $searchPassword;
+    private $defaultRoles;
+    private $defaultSearch;
+
+    /**
+     * @param LdapInterface $ldap
+     * @param string        $baseDn
+     * @param string        $searchDn
+     * @param string        $searchPassword
+     * @param array         $defaultRoles
+     * @param string        $uidKey
+     * @param string        $filter
+     */
+    public function __construct(LdapInterface $ldap, $baseDn, $searchDn = null, $searchPassword = null, array $defaultRoles = array(), $uidKey = 'sAMAccountName', $filter = '({uid_key}={username})')
+    {
+        $this->ldap = $ldap;
+        $this->baseDn = $baseDn;
+        $this->searchDn = $searchDn;
+        $this->searchPassword = $searchPassword;
+        $this->defaultRoles = $defaultRoles;
+        $this->defaultSearch = str_replace('{uid_key}', $uidKey, $filter);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadUserByUsername($username)
+    {
+        try {
+            $this->ldap->bind($this->searchDn, $this->searchPassword);
+            $username = $this->ldap->escape($username, '', LdapInterface::LDAP_ESCAPE_FILTER);
+            $query  = str_replace('{username}', $username, $this->defaultSearch);
+            $search = $this->ldap->find($this->baseDn, $query);
+        } catch (ConnectionException $e) {
+            throw new UsernameNotFoundException(sprintf('User "%s" not found.', $username), 0, $e);
+        }
+
+        if ($search['count'] > 1) {
+            throw new UsernameNotFoundException('More than one user found');
+        }
+
+        $user = $search[0];
+
+        return $this->loadUser($username, $user);
+    }
+
+    public function loadUser($username, $user)
+    {
+        $password = isset($user['userpassword']) ? $user['userpassword'] : null;
+
+        $roles = $this->defaultRoles;
+
+        return new User($username, $password, $roles);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function refreshUser(UserInterface $user)
+    {
+        if (!$user instanceof User) {
+            throw new UnsupportedUserException(sprintf('Instances of "%s" are not supported.', get_class($user)));
+        }
+
+        return new User($user->getUsername(), null, $user->getRoles());
```

Comment:
@csarrazi Dropping the final constraint won't help here. It is not because you define a child class that Symfony will magically use it here. So it is unrelated to the possibility to use a custom user class